jsign-crypto/src/main/java/net/jsign/jca/VenafiSigningService.java

```diff
@@ -0,0 +1,272 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.jsign.jca;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.security.GeneralSecurityException;
+import java.security.Key;
+import java.security.KeyStoreException;
+import java.security.UnrecoverableKeyException;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.cert.Certificate;
+import java.security.cert.CertificateException;
+import java.security.cert.CertificateFactory;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Base64;
+import java.util.LinkedHashMap;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.io.ByteArrayOutputStream;
+import java.math.BigInteger;
+import org.bouncycastle.asn1.ASN1Encodable;
+import org.bouncycastle.asn1.DLSequence;
+import org.bouncycastle.asn1.ASN1Integer;
+import org.bouncycastle.asn1.ASN1OutputStream;
+import org.bouncycastle.asn1.ASN1EncodableVector;
+import org.bouncycastle.asn1.ASN1InputStream;
+import org.bouncycastle.asn1.ASN1ObjectIdentifier;
+import org.bouncycastle.asn1.ASN1OctetString;
+import org.bouncycastle.asn1.ASN1Sequence;
+import org.bouncycastle.asn1.DERSequence;
+import org.bouncycastle.asn1.DEROctetString;
+
+import net.jsign.DigestAlgorithm;
+
+/**
+ * Signing service using the Venafi CodeSign Protect REST Service API.
+ *
+ * @since 7.2
+ * @see <a href="https://docs.venafi.com/Docs/current/TopNav/Content/SDK/CodeSignSDK/cco-SDKc-Code-Signing-SDK-reference.php">Venafi CodeSign Protect REST API</a>
+ */
+public class VenafiSigningService implements SigningService {
+
+    /** Cache of certificates indexed by alias */
+    private final Map<String, Map<String, ?>> certificates = new LinkedHashMap<>();
+
+    /** The API endpoint of the Venafi CodeSign Protect */
+    private final String endpoint;
+
+    private final RESTClient client;
+
+    /** The credentials to authenticate with the service */
+    private final VenafiCredentials credentials;
+
+     /** The name of the Venafi CodeSign Protect certificate KeyId */
+     private String KeyId;
+
+    /** Mapping between Java and Venafi CodeSign Protect signing algorithms */
+    private final Map<String, Integer> algorithmMapping = new HashMap<>();
+    {
+        algorithmMapping.put("SHA256withRSA", 64);
+        algorithmMapping.put("SHA384withRSA", 65);
+        algorithmMapping.put("SHA512withRSA", 66);
+        algorithmMapping.put("SHA256withECDSA", 4164);
+        algorithmMapping.put("SHA384withECDSA", 4165);
+        algorithmMapping.put("SHA512withECDSA", 4166);
+    }
+
+    /* Add ASN.1 DER prefix to MessageDigest */
+
+    public static byte[] addASN1Prefix(DigestAlgorithm digestAlgorithm, byte[] hash) throws IOException {
```

Comment:
This looks like a standard `DigestInfo` structure, it's easier to use the class provided by Bouncy Castle:

    return new DigestInfo(new AlgorithmIdentifier(digestAlgorithm.oid, DERNull.INSTANCE), hash);